### DISCUSSION TITLE
WS Addressing in SOAP Client Request

### DISCUSSION THREAD

**Discussion by Framus** [quarkiverse/quarkus-cxf#999] [Q&A] [answered]
Hello everybody,

I am currently developing a service which acts as a adapter for a legacy SOAP service, therefore I use the quarkus-cxf plugin to generate and implement the SOAP client based on the services wsdl.

So far so good. I already got everything up and running including ws-Security.

Additionally to ws-security, the service uses the ws-addressing `From` field, to obtain an app token.
I can see [here](https://docs.quarkiverse.io/quarkus-cxf/dev/reference/extensions/quarkus-cxf.html) that ws-addressing is supported, but I was not able to find any docs or examples how to use it in scope of a soap **Client**

For ws-security, there is the `WSS4JInInterceptor`. Is there something similar for ws-addressing?

In the end, I need to add something like this in the soap request:

```xml

<soapenv:Header xmlns:wsa="http://www.w3.org/2005/08/addressing">
     <wsa:From><wsa:Address>appToken:’Actual appToken value’</wsa:Address> </wsa:From>
</soapenv:Header>
```

I would appreciate any help or hint where to find documentation about this

Cheers 
Framus

**Comment by ppalaga** [comment]
We do not have anything Quarkus CXF specific at the moment. You need to configure the client in the same way as you would with plain CXF - see e.g. here: https://stackoverflow.com/questions/27419107/apache-cxf-ws-addressing-how-to-set-from-replyto-headers
If the token is constant throughout the whole life of the application, then you can set it at application startup, something like:

```
@ApplicationScoped
public class Init {

    @CXFClient("myClient")
    MyService myService;

    void onStart(@Observes StartupEvent ev) {

        AddressingProperties maps = new AddressingPropertiesImpl();
        maps.setFrom(...);
        ((BindingProvider)myService).getRequestContext()
                .put("javax.xml.ws.addressing.context", maps);

    }
}
```

**Comment by Framus** [comment, marked as answer]
Thank you for the fast answer and the hint.

IT was the right direction. To make the whole thing work I had to do the following:

* Enable the addressing feature by adding `org.apache.cxf.ws.addressing.WSAddressingFeature` to the `quarkus.cxf.client.myClient.features` property
* The context form the CXF example has to be adapted to work with jarkata: 
```((BindingProvider)myService).getRequestContext().put("jakarta.xml.ws.addressing.context", maps);```

When I find the time I will open a PR to maybe extend the documentation here if this is wanted.

So thank you again
Cheers
Framus
